src/recipes/tooling-engine-singleton.md

```diff
@@ -0,0 +1,119 @@
+
+- are Object types
+- are always accessible to GDScript (and other available languages like C#)
+- must be manually registered and unregistered in the `LevelInit::Scene` step
+
+Godot does have _many_ built-in singletons provided internally and exposed to GDScript. You can find a full list
+[here](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#properties)
+
+
+## Defining a Singleton
+
+Defining a singleton is the same as registering a custom class.
+
+```rust
+#[derive(GodotClass)]
+#[class(tool, base=Object)]
+struct MyEditorSingleton {
+    #[base]
+    base: Base<Object>,
+}
+
+#[godot_api]
+impl IObject for MyEditorSingleton {
+    fn init(base: Base<Object>) -> Self {
+        // you may want to omit this by adding "init" to the #[class(...)]
+        Self { base }
+    }
+}
+
+#[godot_api]
+impl MyEditorSingleton {
+    #[func]
+    fn foo(&mut self) {}
+}
+```
+
+
+## Registering a singleton
+
+Registering singletons is done during the `InitLevel::Scene` stage of initialization.
+
+To achieve this, we can customize our init/shutdown routines by overriding ExtensionLibrary trait methods.
+
+```rust
+struct MyExtension;
+
+unsafe impl ExtensionLibrary for MyExtension {
+    fn on_level_init(level: InitLevel) {
+        if level == InitLevel::Scene {
+            // The StringName identifies your singleton and can be used later to access it.
+            Engine::singleton().register_singleton(
+                StringName::from("MyEditorSingleton"),
+                MyEditorSingleton::new_alloc().upcast(),
+            );
+        }
+    }
+
+    fn on_level_deinit(level: InitLevel) {
+        if level == InitLevel::Scene {
+            // Unregistering is needed to avoid memory leaks and warnings, especially for hot reloading.
+            Engine::singleton().unregister_singleton(StringName::from("MyEditorSingleton"));
+        }
+    }
+}
+```
+
+
+## Calling from GDScript
+
+Now that your Engine Singleton is available (and once you've rebuilt and reloaded!) you should be able to access the singleton like so
+
+```godot
+extends Node
+
+func _ready() -> void:
+    MyEditorSingleton.foo()
+
+```
+
+
+## Calling from Rust
+
+You may also want to access your singleton from Rust as well. This can be done through
+[this method](https://godot-rust.github.io/docs/gdext/master/godot/engine/struct.Engine.html#method.get_singleton)
+
+```rust
+godot::engine::Engine::singleton().get_singleton(StringName::from("MyEditorSingleton"));
+```
+
+
+## Singletons and the SceneTree
+
+Singletons cannot safely access the scene tree. At any given moment they may exist without a SceneTree being active.
+While it is technically possible to access the SceneTree through hacky methods, it is **highly recommended** to use a
+custom `EditorPlugin` for this purpose. As you can register a custom "Autoload Singleton" which is crucially different
+from an Engine Singleton in that it is a Node (or child type) instanced on the SceneTree.
```

Comment:
```suggestion
## Singletons and the `SceneTree`

Singletons cannot safely access the scene tree. At any given moment, they may exist without a scene tree being active.
While it is technically possible to access the tree through hacky methods, it is **highly recommended** to use a
custom `EditorPlugin` for this purpose. As you can register a custom "Autoload Singleton" which is crucially different
from an Engine Singleton in that it is a Node (or child type) instanced on the SceneTree.
```

There is one sentence staring with "As", which is unusual. Can you relate it more clearly to the previous part?

Reply:
Still open 🙂